### ISSUE TITLE
Site Editor Template & Global Style Customizations Missing After Atomic Transfer

### ISSUE THREAD

**Issue by mikeicode** [Automattic/wp-calypso#82363]
### Quick summary

We build many sites and customize TT3 on Simple sites, including adding homepage content using the site editor instead of a static page. Occasionally, when the site is transferred to AT, the template changes and the global styles don't transfer.

We have started saving backups of the FSE theme, which can be imported into the AT site as a new theme, but this doesn't resolve the actual issue.

We tested this p1690989410854689-slack-C0160HSMDQV, and it's hard to recreate. However, it has happened in multiple instances, and when it does occur, it's a serious issue since the entire homepage and all customizations are lost.

Recent example: p1694804177869939-slack-C03JUG3KMBN

I'm going to mark this "Yes, easy to implement", but the reality is that if you didn't happen to export the Site Editor "Export > Download your theme with updated templates and styles." theme, it's not possible to fix without reverting the site. 

### Steps to reproduce

- Use the site editor to edit templates and global styles on TT3 on a simple site.
- Push site AT

### What you expected to happen

All TT3 customizations should remain. Simple site:

![Screen Shot on 2023-08-02 at 11_03_03](https://github.com/Automattic/wp-calypso/assets/10353785/c08c7027-0a83-47ed-9dd4-37eb92d2d3f0)

### What actually happened

None of the changes transferred to AT:

![Screen Shot on 2023-08-02 at 11-07-32](https://github.com/Automattic/wp-calypso/assets/10353785/81302933-d466-4c12-9f47-6659fdcfc86f)

### Impact

Some (< 50%)

### Available workarounds?

Yes, easy to implement

### Platform (Simple and/or Atomic)

Atomic

### Logs or notes

We've only run into this on Twenty Twenty-Three, but that is our most commonly used theme.

**Comment by cuemarie** [comment]
📌 REPRODUCTION RESULTS
- Tested on Simple-> AT - Could Not Replicate

📌 FINDINGS/SCREENSHOTS/VIDEO
I was not able to repro, but as noted above, **this is expected** as the issue is intermittent and not always happening.  
I customized the Blog Home template of TT3 and added Global Style changes (colors), however these remained after AT-ing the site.

📌 ACTIONS
- Triaged

**Comment by metabreakr** [comment]
#81593 might be related.

**Comment by mikeicode** [comment]
This happened again to a BBE customer. pdEkSe-3VB-p2#comment-2443

**Comment by mikeicode** [comment]
Another customer ran into this. We didn't have the FSE export, so we had to revert the site. The global styles were missing, so those had to be added back. Then the site could go AT and have the FSE theme installed. pdEkSe-3MF-p2#comment-2553

**Comment by mikeicode** [comment]
Ran into this again when a customer pushed installed a plugin. I installed the FSE theme to "resolve".

pdEkSe-5U0-p2#comment-2630
p1698413378524169-slack-C0160HSMDQV

**Comment by marissaexplores** [comment]
Here is another instance where the customer went atomic and lost the template layouts and global styles to TT3. We restored it with a saved FSE export.

pdEkSe-5oC-p2#comment-2637

**Comment by cuemarie** [comment]
📌 ACTIONS
- Increasing priority & will try to get some eyes on this
- Raised for awareness here: p1698700952206589-slack-C01LXBU1N21

**Comment by mikeicode** [comment]
For troubleshooting, here are a couple of examples of sites with the issue. These are my test sites; feel free to make any changes.

https://testdifm249412112.wpcomstaging.com - The site from the images in this report.

https://testbbemike67.wpcomstaging.com - There is a video showing the entire process resulting in the issue on this site here:

 p1690993729222909/1690989410.854689-slack-C0160HSMDQV

We do often import an XML into the simple site at the start of our building process, perhaps this is somehow related. The XML includes the site editor template changes, and usually global styles are also imported. An example XML is available here: 

p1690993825016369/1690989410.854689-slack-C0160HSMDQV

**Comment by mikeicode** [comment]
Ran into this on another site, details and screenshots: pdEkSe-5Rb-p2#comment-2656.

**Comment by inaikem** [comment]
Testing: pdEkSe-5oC-p2

**Comment by mikeicode** [comment]
Ran into this again: pdEkSe-6sI-p2#comment-2755

I imported the XML from [this](https://href.li/?https://expressdemo12.wordpress.com/) simple site demo, then built the site on simple. I pushed the site AT from hosting configuration, all changes were lost. I imported the FSE theme and added the custom CSS back to “resolve”.

**Comment by BogdanUngureanu** [comment]
Hey @mikeicode, I've exported the XML from that demo site and imported it on a new site. However, I haven't managed to reproduce the issue. 

The steps I've followed:
 - Exported the site from the demo site
 - Created a new site on my account with a Business subscription
 - Imported the content on the new site.
 - Noticed that the Advanced CSS configuration wasn't imported
 - Noticed that the UI elements were green instead of looking like the demo site
 - Went to Settings > Hosting configuration
 - Migrated the site to AT
 - The site looks the same as it did on the Simple Site (with the green UI elements).

Can you share the export you've made of the site? You could send it me on Slack or in the T-Rex channel.

**Comment by mikeicode** [comment]
@BogdanUngureanu Thanks for taking a look, I'll send the XML to you in Slack.

You mentioned:

>Noticed that the Advanced CSS configuration wasn't imported

I've observed that custom CSS never imports with XML import.

> Noticed that the UI elements were green instead of looking like the demo site

This is a separate, frustrating issue, which I tested extensively earlier this year. Slack: p1678187168952379-slack-C01A60HCGUA

The issue is that there is no consistency to if XML imports will import the site editor template change or the global style changes. I’ve found that on simple sites the template changes are usually  imported, it’s a 50/50 if the global style changes will be imported. On Atomic sites, there is an even lower chance that either will be imported. 

Currently, our team just deals with this issue when building sites.

Returning to the simple to Atomic issue, this is a more serious problem since it can happen after a project is completed, when the customer pushes their site Atomic.

After you try the import using the XML I provided, if the global styles aren't visible upon import, try it on a different new site. As previously mentioned, the results are inconsistent. Once you achieve an import with the correct styles, please make additional changes to the homepage template and global styles, as would be done in a real build. Then push the site AT from hosting configuration. If the site appears correct after the upgrade, repeat the process with another simple site until you run into the issue. This bug is also inconsistent. I built 2 sites yesterday for customers using this demo https://expressdemo12.wordpress.com, and the process mentioned above. One lost all customizations after being pushed AT, while the other didn't.

**Comment by mikeicode** [comment]
Another: pdEkSe-6Cd-p2#comment-2839

I imported the XML from [this](https://href.li/?https://expressdemo12.wordpress.com/) simple site demo, then built the site on simple. I pushed the site AT from hosting configuration, all changes were lost. I imported the FSE theme and added the custom CSS back to “resolve”.

**Comment by BogdanUngureanu** [comment]
I tried again to reproduce it on a new site based on an export I got from Mike. Unfortunately, I tried it on a few sites without luck; I wasn't able to reproduce it. However, I did notice that the imported site looked slightly different than the demo one:
- Instead of the blue elements, I'm getting green elements.
- Some CSS was missing on the test site; because of this, the header bar had a top margin.

Keep in mind that these differences were seen on a Simple site, before I've migrated to Atomic and are problem of the import process.

@Automattic/t-rex Do you have some free time to try to reproduce this issue?

**Comment by mikeicode** [comment]
@BogdanUngureanu 

> Instead of the blue elements, I'm getting green elements.
> Some CSS was missing on the test site; because of this, the header bar had a top margin.

Please see my [previous comment ](https://github.com/Automattic/wp-calypso/issues/82363#issuecomment-1806280330)above, where I discussed both of these. TL;DR - CSS never gets imported and the color change (global styles occasionally don't get imported) is a separate issue:

https://github.com/Automattic/wp-calypso/issues/81849

**Comment by mikeicode** [comment]
Another: https://wp.me/pdEkSe-6E0

I imported the XML from [this](https://href.li/?https://expressdemo12.wordpress.com/) simple site demo, then built the site on simple. I pushed the site AT from hosting configuration, all changes were lost. I imported the FSE theme and added the custom CSS back to “resolve”.

**Comment by mikeicode** [comment]
Another, screenshots: pdEkSe-6Bt-p2#comment-2876

**Comment by dsas** [comment]
@mikeicode can you add me as an admin to your test site  testdifm249412112.wordpress.com ?

I've tried a bunch of stuff and haven't been able to repro myself yet.

If the custom CSS thing is reproducible, then maybe if we start working on that we'll get to an insight that helps explain the rest of it? ☹️

**Comment by mikeicode** [comment]
@dsas I added you as an admin on https://testdifm249412112.wpcomstaging.com

> If the custom CSS thing is reproducible, then maybe if we start working on that we'll get to an insight that helps explain the rest of it?

I'm not sure if custom CSS is _supposed_ to be imported with an XML. It would be nice if it was of course.

Perhaps you could try looking at this issue ( https://github.com/Automattic/wp-calypso/issues/81849 ), which is more of a pain for us, and could potentially be related. That issue is what @BogdanUngureanu mentioned above:

> Instead of the blue elements, I'm getting green elements.

**TL;DR of that issue** - When importing an XML file that includes FSE template changes and global styles (such as the one from this current issue) into a simple site, the FSE changes usually import successfully. However, the import of global styles is less reliable, with about a 50% success rate. When importing the same XML file into an Atomic site, it is common for neither the FSE nor the global styles to import successfully.

**Comment by dsas** [comment]
> I'm not sure if custom CSS is supposed to be imported with an XML. It would be nice if it was of course.

I haven't worked with WP importers and exporters since before FSE was a thing but custom css is just stored as a post, I'd expect it to work myself.

----

Looking at the database for testdifm249412112.wordpress.com I can see multiple global styles posts:

 * wp-global-styles-pub%2fzoologist
 * wp-global-styles-pub%2ftwentytwentythree
 * wp-global-styles-pub%2ftwentytwentythree-2
 * wp-global-styles-pub%2ftwentytwentythree-2-2
 * wp-global-styles-premium%2fyuga
 * wp-global-styles-twentytwentythree

The active theme is `twentytwentythree` It appears that the last post listed is the one being used - the style definitions contained are empty.

It looks like the second post - `wp-global-styles-pub%2ftwentytwentythree` is the one that you are expecting to see, it defines a primary colour of #7cb7af and has lots of other changes.

The third and fourth posts - `wp-global-styles-pub%2ftwentytwentythree-2` and `-2` are more minimal, and change the primary colour to be brown.

If I create an atomic site, change the colours and check the database it's creating a global styles post called `wp-global-styles-twentytwentythree`. It works.

If I create a simple site, change the colours and then go atomic, it's just creating a global styles post called `wp-global-styles-pub%2ftwentytwentythree`. It works.

Notice that the `-pub` is there when I start from a simple site, and not when I'm from atomic. I feel this is very related to the problem, but I haven't dug any further yet.

**Comment by dsas** [comment]
Looking at [how Gutenberg gets the global styles data](https://github.com/WordPress/gutenberg/blob/ad88b6b9daaae5ac8a737102c0b586163989da56/lib/class-wp-theme-json-resolver-gutenberg.php#L483) in `get_user_data_from_wp_global_styles` it first runs this SQL:

`SELECT t.term_id
FROM wp_terms AS t INNER JOIN wp_term_taxonomy AS tt ON t.term_id = tt.term_id
WHERE tt.taxonomy IN ('wp_theme') AND t.name IN ('twentytwentythree') LIMIT 1`

There's no order by, so it's pot-luck as to whether it returns the term_id for the term_taxonomy_id associated with the pub/twentytwentythree GS post or the twentytwentythree GS post. It then uses that ID to get the GS post.

Depending upon how the function is called, if it doesn't find a relevant post, it may insert a blank global styles post (like the twentytwentythree one).

I'd guess that is why this error isn't always reproducible. Something happens during some atomic transfers that results in a blank GS post being created, furthermore some or all of these sites always end up selecting the "blank GS taxonomy" rather than the populated one (likely MySQL returns the first one it finds on disk/in index or something). 

Complete speculation, but perhaps it's as simple as someone tries to access the site during the import. Perhaps exacerbated if TT3 is the default theme as the theme might exist and be active before the database has finished importing?

~~Edit: This would still not explain the templates apparently not being transferred~~

## Next steps
 1. Find out if this is perhaps caused by something we're doing during the transfer
 2. Maybe pick the brains of atomic people

**Comment by dsas** [comment]
@mikeicode does testdifm249412112.wordpress.com have missing templates or just missing global styles? If you have a test atomic site with missing templates, could you add me as an admin to that too please?

**Comment by mikeicode** [comment]
@dsas Yes, that site is missing the entire homepage, which was created in the editor. If you view the images at the start of this issue, notice how on simple the template was called "Home", but after pushing the site AT the main page in the editor is "Blog Home", and I don't see a "Home" template listed in the [templates](https://testdifm249412112.wpcomstaging.com/wp-admin/site-editor.php?path=%2Fwp_template%2Fall).

**Comment by dsas** [comment]
Thanks @mikeicode, sorry you did already explain that clearly.

It looks like so far as selecting templates goes it's a similar thing - there are two term_ids and it's using the same incorrect one when looking for the template.

**Comment by mikeicode** [comment]
The primary issue from this bug report, which is content loss when pushing AT, occurred on another customer's site. We resolved it again with the FSE theme export.

p1702060156109449-slack-C041ZS8R1N0

**Comment by mikeicode** [comment]
Another on a customer's site. pdEkSe-7bQ-p2#comment-3131

**Comment by edequalsawesome** [comment]
Another report here: #7510740-zd

**Comment by mikeicode** [comment]
I tested this and ran into the same issue.

* Create a new simple site on Creator plan
* Create new XML export from simple site demo (https://expressdemo12.wordpress.com)
* Import XML into new simple site, everything worked well.
* Make changes in the site editor. Global style changes, block changes
* Push site AT from Settings > Hosting Configuration (https://mikeicode82.wpcomstaging.com)
* Check site editor, it's default TT3. The homepage content and customizations are missing.

Site on simple:

![Screenshot 2024-01-25 at 11 21 44 AM](https://github.com/Automattic/wp-calypso/assets/10353785/9d0344c1-4e7b-45fa-b40a-a9f18b3a7de1)

Site on AT:

![Screenshot 2024-01-25 at 11 25 30 AM](https://github.com/Automattic/wp-calypso/assets/10353785/11264e9f-34b1-4d53-bbfe-a5ffa6ae575a)

**Comment by pauloeaquino** [comment]
Another report here pdEkSe-7eO#comment-3434

**Comment by ouikhuan** [comment]
@mikeicode , we've shipped two diff to address the following issue:
- Remove the original terms after we add the prefix on simple site. (D136207-code)
- Remove all the prefix in export file including inner blocks. (D136355-code)

Could you verify this from your end and see if it works now? Thank you!

**Comment by mikeicode** [comment]
@ouikhuan I tested this on a test site, and it worked well. More importantly, I tested it on an actual build for a customer, and everything was correct after the AT transfer. I'm going to mark this issue as resolved and will reopen it if we encounter this problem again, though hopefully, we won't need to. Thanks for your work on this!
